### PR TITLE
Fix: Expense - Not here page shows up briefly when deleting the expense

### PR DESCRIPTION
**Pull request by wildan-m**
### Explanation of Change

This PR will not here page briefly appears when deleting the expense by navigating to target url then perform actual delete when the report details unmounted.

### Fixed Issues

$ https://github.com/Expensify/App/issues/45576
PROPOSAL: https://github.com/Expensify/App/issues/45576#issuecomment-2309556187

### Tests

1. Open App
2. Go to DM.
3. Submit two expenses to the user.
4. Open any of the transaction thread.
5. Tap on the report header.
6. Tap Delete expense..
7. Delete the expense.
8. Verify Not here page will not show up when deleting the expense.
9. Verify [Deleted Expense] text not shown in report details page
- [X] Verify that no errors appear in the JS console

### Offline tests

Same as test
### QA Steps

Same as test
- [X] Verify that no errors appear in the JS console

### PR Author Checklist

- [X] I linked the correct issue in the `### Fixed Issues` section above
- [X] I wrote clear testing steps that cover the changes made in this PR
    - [X] I added steps for local testing in the `Tests` section
    - [X] I added steps for the expected offline behavior in the `Offline steps` section
    - [X] I added steps for Staging and/or Production testing in the `QA steps` section
    - [X] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [X] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
    - [X] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [X] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [X] I ran the tests on **all platforms** & verified they passed on:
    - [X] Android: Native
    - [X] Android: mWeb Chrome
    - [X] iOS: Native
    - [X] iOS: mWeb Safari
    - [X] MacOS: Chrome / Safari
    - [X] MacOS: Desktop
- [X] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [X] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [X] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [X] I verified that comments were added to code that is not self explanatory
    - [X] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [X] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
      - [X] If any non-english text was added/modified, I verified the translation was requested/reviewed in #expensify-open-source and it was approved by an internal Expensify engineer. Link to Slack message:
    - [X] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [X] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is either coming verbatim from figma or has been approved by marketing (in order to get marketing approval, ask the Bug Zero team member to add the Waiting for copy label to the issue)
    - [X] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [X] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [X] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [X] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [X] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [X] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [X] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [X] I verified that if a function's arguments changed that all usages have also been updated correctly
- [X] If any new file was added I verified that:
    - [X] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [X] If a new CSS style is added I verified that:
    - [X] A similar style doesn't already exist
    - [X] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/utils/index.ts) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(theme.componentBG)`)
- [X] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [X] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [X] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [X] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [X] If the PR modifies the UI (e.g. new buttons, new UI components, changing the padding/spacing/sizing, moving components, etc) or modifies the form input styles:
    - [X] I verified that all the inputs inside a form are aligned with each other.
    - [X] I added `Design` label and/or tagged `@Expensify/design` so the design team can review the changes.
- [X] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [X] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.

### Screenshots/Videos
<details>
<summary>Android: Native</summary>

https://github.com/user-attachments/assets/d19a4fed-3024-44a6-8ccc-b6325749df4a

</details>

<details>
<summary>Android: mWeb Chrome</summary>

https://github.com/user-attachments/assets/9102a78d-2d3c-4264-a707-16475c80a535

</details>

<details>
<summary>iOS: Native</summary>

https://github.com/user-attachments/assets/a3599a80-78f5-48fd-a7a6-a9d30a916909

</details>

<details>
<summary>iOS: mWeb Safari</summary>

https://github.com/user-attachments/assets/25ff0228-3140-45ce-82d9-68eaa7425b5c

</details>

<details>
<summary>MacOS: Chrome / Safari</summary>

https://github.com/user-attachments/assets/eb0fd82b-2afc-4b0f-bfa0-ca7f59643479

</details>

<details>
<summary>MacOS: Desktop</summary>

https://github.com/user-attachments/assets/855184ac-a02e-4e9a-ac16-c0ffdb191f55

</details>

### CODE REVIEW

File: src/ONYXKEYS.ts [hunk 1]
```diff
@@ -219,6 +219,9 @@ const ONYXKEYS = {
     /** The NVP containing all information related to educational tooltip in workspace chat */
     NVP_WORKSPACE_TOOLTIP: 'workspaceTooltip',
 
+    /**  The NVP contain url to back after deleting transaction */
```
Comment (luacmartins) [inline src/ONYXKEYS.ts:222]:
```suggestion
    /**  The NVP containing the target url to navigate to when deleting a transaction */
```

File: src/libs/actions/IOU.ts [hunk 2]
```diff
@@ -1740,9 +1740,18 @@ function getDeleteTrackExpenseInformation(
...
+                value: {
+                    reportID: null,
+                    stateNum: CONST.REPORT.STATE_NUM.APPROVED,
+                    statusNum: CONST.REPORT.STATUS_NUM.CLOSED,
+                    participants: {
+                        [userAccountID]: {
+                            notificationPreference: CONST.REPORT.NOTIFICATION_PREFERENCE.HIDDEN,
+                        },
+                    },
+                },
```
Comment (luacmartins) [inline src/libs/actions/IOU.ts:1754]: Why are we changing this? When the transaction is deleted we should delete the transaction thread, no?

File: src/libs/actions/IOU.ts [hunk 3]
```diff
@@ -1782,6 +1791,17 @@ function getDeleteTrackExpenseInformation(
...
+        successData.push({
+            onyxMethod: Onyx.METHOD.MERGE,
+            key: `${ONYXKEYS.COLLECTION.REPORT}${transactionThreadID}`,
+            value: Object.keys(transactionThread).reduce<Record<string, null>>((acc, key) => {
+                acc[key] = null;
+                return acc;
+            }, {}),
+        });
+    }
+
```
Comment (luacmartins) [inline src/libs/actions/IOU.ts:1804]: Why is this needed?

File: src/libs/actions/IOU.ts [hunk 4]
```diff
@@ -5497,10 +5503,59 @@ function prepareToCleanUpMoneyRequest(transactionID: string, reportAction: OnyxT
...
 }
 
+/**
+ * Calculate the URL to navigate to after a money request deletion
+ * @param transactionID - The ID of the money request being deleted
+ * @param reportAction - The report action associated with the money request
+ * @param isSingleTransactionView - whether we are in the transaction thread report
+ * @returns The URL to navigate to
+ */
+function getNavigationUrlAfterMoneyRequestDelete(transactionID: string, reportAction: OnyxTypes.ReportAction, isSingleTransactionView = false): string | undefined {
```
Comment (luacmartins) [inline src/libs/actions/IOU.ts:5516]:
```suggestion
function getNavigationUrlOnMoneyRequestDelete(transactionID: string, reportAction: OnyxTypes.ReportAction, isSingleTransactionView = false): string | undefined {
```

File: src/libs/actions/IOU.ts [hunk 5]
```diff
@@ -5679,9 +5735,18 @@ function deleteMoneyRequest(transactionID: string, reportAction: OnyxTypes.Repor
...
+                value: {
+                    reportID: null,
+                    stateNum: CONST.REPORT.STATE_NUM.APPROVED,
+                    statusNum: CONST.REPORT.STATUS_NUM.CLOSED,
+                    participants: {
+                        [userAccountID]: {
+                            notificationPreference: CONST.REPORT.NOTIFICATION_PREFERENCE.HIDDEN,
+                        },
+                    },
+                },
```
Comment (luacmartins) [inline src/libs/actions/IOU.ts:5749]: Same here

File: src/libs/actions/IOU.ts [hunk 6]
```diff
@@ -5772,6 +5837,17 @@ function deleteMoneyRequest(transactionID: string, reportAction: OnyxTypes.Repor
...
+        successData.push({
+            onyxMethod: Onyx.METHOD.MERGE,
+            key: `${ONYXKEYS.COLLECTION.REPORT}${transactionThreadID}`,
+            value: Object.keys(transactionThread).reduce<Record<string, null>>((acc, key) => {
+                acc[key] = null;
+                return acc;
+            }, {}),
+        });
+    }
+
```
Comment (luacmartins) [inline src/libs/actions/IOU.ts:5850]: and here

File: src/libs/actions/IOU.ts [hunk 7]
```diff
@@ -5898,10 +5977,11 @@ function deleteTrackExpense(chatReportID: string, transactionID: string, reportA
...
     // STEP 7: Navigate the user depending on which page they are on and which resources were deleted
-    if (isSingleTransactionView && shouldDeleteTransactionThread) {
-        // Pop the deleted report screen before navigating. This prevents navigating to the Concierge chat due to the missing report.
-        return ROUTES.REPORT_WITH_ID.getRoute(chatReport?.reportID ?? '-1');
-    }
+    return urlToNavigateBack;
+    // if (isSingleTransactionView && shouldDeleteTransactionThread) {
+    //     // Pop the deleted report screen before navigating. This prevents navigating to the Concierge chat due to the missing report.
+    //     return ROUTES.REPORT_WITH_ID.getRoute(chatReport?.reportID ?? '-1');
+    // }
```
Comment (luacmartins) [inline src/libs/actions/IOU.ts:5984]: It seems like we can remove this

File: src/libs/actions/Task.ts [hunk 8]
```diff
@@ -959,6 +959,36 @@ function getParentReport(report: OnyxEntry<OnyxTypes.Report>): OnyxEntry<OnyxTyp
     return ReportConnection.getAllReports()?.[`${ONYXKEYS.COLLECTION.REPORT}${report.parentReportID}`];
 }
 
+/**
+ * Calculate the URL to navigate to after a task deletion
+ * @param report - The task report being deleted
+ * @returns The URL to navigate to
+ */
+function getNavigationUrlAfterTaskDelete(report: OnyxEntry<OnyxTypes.Report>): string | undefined {
```
Comment (luacmartins) [inline src/libs/actions/Task.ts:967]:
```suggestion
function getNavigationUrlOnTaskDelete(report: OnyxEntry<OnyxTypes.Report>): string | undefined {
```

File: src/pages/ReportDetailsPage.tsx [hunk 9]
```diff
@@ -794,31 +794,73 @@ function ReportDetailsPage({policies, report, route}: ReportDetailsPageProps) {
...
+
+        if (isTrackExpense) {
+            IOU.deleteTrackExpense(moneyRequestReport?.reportID ?? '', iouTransactionID, requestParentReportAction, isSingleTransactionView);
         } else {
-            navigateBackToAfterDelete.current = IOU.deleteMoneyRequest(iouTransactionID, requestParentReportAction, isSingleTransactionView);
+            IOU.deleteMoneyRequest(iouTransactionID, requestParentReportAction, isSingleTransactionView);
         }
+    }, [caseID, iouTransactionID, isSingleTransactionView, moneyRequestReport?.reportID, report, requestParentReportAction]);
+
+    // A flag to indicate whether the user choose to delete the transaction or not
```
Comment (luacmartins) [inline src/pages/ReportDetailsPage.tsx:816]:
```suggestion
    // A flag to indicate whether the user chose to delete the transaction or not
```

File: src/pages/ReportDetailsPage.tsx [hunk 10]
```diff
@@ -794,31 +794,73 @@ function ReportDetailsPage({policies, report, route}: ReportDetailsPageProps) {
...
+            // Perform the actual deletion after the details page is unmounted. This prevents the [Deleted ...] text from briefly appearing when dismissing the modal.
+            if (!isTransactionDeleted.current) {
+                return;
+            }
+
+            deleteTransaction();
+        };
+    }, [deleteTransaction]);
+
+    // Where to go back after deleting the transaction and its report.
```
Comment (luacmartins) [inline src/pages/ReportDetailsPage.tsx:830]:
```suggestion
    // Where to navigate back to after deleting the transaction and its report.
```

**Review by luacmartins — CHANGES_REQUESTED** [review]
Left a few comments. TS is also failing

**Comment by wildan-m** [comment]
> TS is also failing

@luacmartins @alitoshmatov do you know why this error shown? seems not directly related with the PR

image

File: src/libs/actions/IOU.ts (hunk 2 above)
Comment (wildan-m) [reply src/libs/actions/IOU.ts:1758]: @luacmartins We switched from using `SET` to `MERGE` to prevent the appearance of the not found page due to platforms updating the value too quickly. This change has already been made for [`leaveRoom`](https://github.com/Expensify/App/blob/7cb72604f0129e2771ac72bd83d1b33e9b80d594/src/libs/actions/Report.ts#L2971-L2980) and [`leaveGroupChat`](https://github.com/Expensify/App/blob/7cb72604f0129e2771ac72bd83d1b33e9b80d594/src/libs/actions/Report.ts#L2898-L2905) to address a similar issue https://github.com/Expensify/App/issues/49468.

File: src/libs/actions/IOU.ts (hunk 3 above)
Comment (wildan-m) [reply src/libs/actions/IOU.ts:1808]: @luacmartins Ensure that any remaining data is removed upon successful completion, even if the server sends a report removal response. This is done to prevent the removal update from lingering in the `applyHTTPSOnyxUpdates` function. https://github.com/Expensify/App/pull/50846#discussion_r1802366119

File: src/libs/actions/IOU.ts (hunk 2 above)
Comment (luacmartins) [reply src/libs/actions/IOU.ts:1759]: Thanks for clarifying. I think that makes sense, it just feels a bit odd to do that. Let's add a comment explaining why we're doing it so people don't try to revert this change

File: src/libs/actions/IOU.ts (hunk 3 above)
Comment (luacmartins) [reply src/libs/actions/IOU.ts:1809]: Same here, let's add a comment about it. We should add a comment for all instances to make sure people don't revert this

**Review by luacmartins — DISMISSED** [review]
LGTM. @alitoshmatov all yours

File: src/libs/actions/IOU.ts (hunk 2 above)
Comment (wildan-m) [reply src/libs/actions/IOU.ts:1801]: done

File: src/libs/actions/IOU.ts (hunk 3 above)
Comment (wildan-m) [reply src/libs/actions/IOU.ts:1853]: done

File: src/libs/actions/IOU.ts (hunk 5 above)
Comment (wildan-m) [reply src/libs/actions/IOU.ts:5789]: done

File: src/libs/actions/IOU.ts (hunk 6 above)
Comment (wildan-m) [reply src/libs/actions/IOU.ts:5899]: done

**Review by luacmartins — APPROVED** [review]
Thanks for addressing the change requests. LGTM. @alitoshmatov all yours

**Comment by alitoshmatov** [comment]
@wildan-m 
When expense has a comment, navigation is behaving a little bit different. When you delete the expense:
1. Modal is closing
2. Existing expense page is shown
3. Then Expense report is shown

I think this is too much, can't we just navigate to the final page instantly?


https://github.com/user-attachments/assets/d847b408-d81f-43dc-bf9e-e3455c088032

https://github.com/user-attachments/assets/eee8eccc-f2ab-439a-b4d5-8e3d0533648d

**Comment by wildan-m** [comment]
@alitoshmatov Thanks for your feedback I've got this loading error in dev and staging 

Seems BE?
`Unexpected error deleting this expense. Please try again later`

https://github.com/user-attachments/assets/88bb7c2e-537e-43eb-ba9d-aba0972c6c11

I'll try to fix the issue after that BE issue resolved

**Comment by luacmartins** [comment]
@wildan-m is that consistently reproducible? Do you have a requestID for a failing request?

**Comment by wildan-m** [comment]
@luacmartins yes, it consistently reproduced, I think it's FE.

Asking if anyone else has experienced the same.
https://expensify.slack.com/archives/C01GTK53T8Q/p1732632553483729

**Comment by wildan-m** [comment]
@alitoshmatov can you retest with this new merged main? I can't reproduce [your mentioned issue](https://github.com/Expensify/App/pull/52740#issuecomment-2499886607) I can only reproduce [the loading issue](https://github.com/Expensify/App/pull/52740#issuecomment-2499918050) which is also occurred in staging.

**Comment by wildan-m** [comment]
> @wildan-m When expense has a comment, navigation is behaving a little bit different. When you delete the expense:
> 
> 1. Modal is closing
> 2. Existing expense page is shown
> 3. Then Expense report is shown
> 


@alitoshmatov actually that behavior is current main behavior, we can't see that clearly because it's deleting and navigation fast. If we throttle the CPU, we'll see it better


https://github.com/user-attachments/assets/0d3acf1f-8e34-47e7-9311-4bfd399bc5ee

Should we change that behavior?

**Comment by luacmartins** [comment]
>Also, I've noticed that isDeletedParentAction returned by DeleteMoneyRequest API is false, where it expected to be true

I found the backend bug for this. I'll work on a fix.

**Comment by luacmartins** [comment]
>IIRC the expected behavior if there is a comment when deleting expense is -- it should navigate to the existing expense page, not the parent report. @luacmartins do you know if the expected behavior changes?

It seems like the behavior has changed as you mentioned.

**Comment by alitoshmatov** [comment]
> @alitoshmatov actually that behavior is current staging behavior, we can't see that clearly because it's deleting and navigation fast. If we throttle the CPU, we'll see it better

I see

>  if there is a comment when deleting expense is -- it should navigate to the existing expense page, not the parent report

So we decided the behavior should be as you said, right?

**Comment by luacmartins** [comment]
That's my understanding

**Comment by wildan-m** [comment]
@luacmartins @alitoshmatov should we fix the behavior that accidentally changed by this PR https://github.com/Expensify/App/pull/45563 here?

**Comment by luacmartins** [comment]
I don't think we should change that as part of this PR. If we decide to change it, we should do so in a separate one.

**Comment by wildan-m** [comment]
> I don't think we should change that as part of this PR. If we decide to change it, we should do so in a separate one.

@luacmartins @alitoshmatov Then should [this behavior](https://github.com/Expensify/App/pull/52740#issuecomment-2499886607)  be accepted, since the navigation successfully takes us to the correct screen, but then the useEffect in ReportScreen re-navigates us to the parent report. Right?

**Comment by luacmartins** [comment]
Yes, I agree.

**Review by luacmartins — APPROVED** [review]

**Comment by alitoshmatov** [comment]
## Reviewer Checklist

- [x] I have verified the author checklist is complete (all boxes are checked off).
- [x] I verified the correct issue is linked in the `### Fixed Issues` section above
- [x] I verified testing steps are clear and they cover the changes made in this PR
    - [x] I verified the steps for local testing are in the `Tests` section
    - [x] I verified the steps for Staging and/or Production testing are in the `QA steps` section
    - [x] I verified the steps cover any possible failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
- [x] I checked that screenshots or videos are included for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I verified tests pass on **all platforms** & I tested again on:
    - [x] Android: Native
    - [x] Android: mWeb Chrome
    - [x] iOS: Native
    - [x] iOS: mWeb Safari
    - [x] MacOS: Chrome / Safari
    - [x] MacOS: Desktop
- [x] If there are any errors in the console that are unrelated to this PR, I either fixed them (preferred) or linked to where I reported them in Slack
- [x] I verified proper code patterns were followed (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`).
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is either coming verbatim from figma or has been approved by marketing (in order to get marketing approval, ask the Bug Zero team member to add the Waiting for copy label to the issue)
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I verified that this PR follows the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I verified other components that can be impacted by these changes have been tested, and I retested again (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` have been tested & I retested again)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.ts or at the top of the file that uses the constant) are defined as such
- [x] If a new component is created I verified that:
    - [x] A similar component doesn't exist in the codebase
    - [x] All props are defined accurately and each prop has a `/** comment above it */`
    - [x] The file is named correctly
    - [x] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [x] The only data being stored in the state is data necessary for rendering and nothing else
    - [x] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [x] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [x] All JSX used for rendering exists in the render method
    - [x] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/utils/index.ts) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(theme.componentBG`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [x] If the PR modifies the UI (e.g. new buttons, new UI components, changing the padding/spacing/sizing, moving components, etc) or modifies the form input styles:
    - [x] I verified that all the inputs inside a form are aligned with each other.
    - [x] I added `Design` label and/or tagged `@Expensify/design` so the design team can review the changes.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] For any bug fix or new feature in this PR, I verified that sufficient [unit tests](https://github.com/Expensify/App/blob/main/tests/README.md) are included to prevent regressions in this flow.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.
- [x] I have checked off every checkbox in the PR reviewer checklist, including those that don't apply to this PR.

### Screenshots/Videos
<details>
<summary>Android: Native</summary>


https://github.com/user-attachments/assets/ccef5ec9-8deb-47a0-8b0f-86da812f674b

</details>

<details>
<summary>Android: mWeb Chrome</summary>



</details>

<details>
<summary>iOS: Native</summary>



https://github.com/user-attachments/assets/a5cda677-3ccd-4c63-976e-8c43e81cd49f

</details>

<details>
<summary>iOS: mWeb Safari</summary>



https://github.com/user-attachments/assets/28895c2f-e26c-470f-af50-8558b3eb40ee

</details>

<details>
<summary>MacOS: Chrome / Safari</summary>

https://github.com/user-attachments/assets/7f3a6f9d-cc93-471c-97d4-ec4a6614f3d8


</details>

<details>
<summary>MacOS: Desktop</summary>

https://github.com/user-attachments/assets/be15d6ab-59e1-4276-aa65-385ce5220b64



</details>

**Review by alitoshmatov — APPROVED** [review]
LGTM!

**Review by luacmartins — APPROVED** [review]

**Comment by wildan-m** [comment]
@luacmartins @alitoshmatov I just remember that there is [a feedback](https://github.com/Expensify/App/pull/50846#discussion_r1864559285) on successData regarding similar pattern we use here. To avoid similar issue should we patch this PR a bit? we'll make the success data to null instead. Here is the patch https://github.com/Expensify/App/pull/53616

File: src/pages/ReportDetailsPage.tsx [hunk 11]
```diff
@@ -797,31 +797,73 @@ function ReportDetailsPage({policies, report, route, reportMetadata}: ReportDeta
...
+            deleteTransaction();
+        };
+    }, [deleteTransaction]);
+
+    // Where to navigate back to after deleting the transaction and its report.
+    const navigateToTargetUrl = useCallback(() => {
+        let urlToNavigateBack: string | undefined;
+
+        if (!isTransactionDeleted.current) {
+            if (caseID === CASES.DEFAULT) {
```
Comment (ikevin127) [inline src/pages/ReportDetailsPage.tsx:838]: The key issue with this new logic is that even when `isTransactionDeleted.current` is `false` (delete confirmation modal Cancel case), it still proceeds to call `Task.getNavigationUrlOnTaskDelete` and navigate to parent report when modal closes because it's passed to `onModalHide={navigateToTargetUrl}`. This triggers navigation back to parent report when user will press `Cancel` on the delete confirmation modal when attempting to delete a task report (see before / after videos in PR mentioned below).

The old `deleteTransaction` code had a simpler check that would only navigate if there was an explicit `navigateBackToAfterDelete.current` value.

We decided to tackle this in PR https://github.com/Expensify/App/pull/53408 since this PR fixed issue https://github.com/Expensify/App/issues/53301, which the other PR was supposed to fix but this one merged before.

File: src/pages/ReportDetailsPage.tsx [hunk 12]
```diff
@@ -797,31 +797,73 @@ function ReportDetailsPage({policies, report, route, reportMetadata}: ReportDeta
...
+    useEffect(() => {
+        return () => {
+            // Perform the actual deletion after the details page is unmounted. This prevents the [Deleted ...] text from briefly appearing when dismissing the modal.
+            if (!isTransactionDeleted.current) {
+                return;
+            }
+
+            deleteTransaction();
+        };
+    }, [deleteTransaction]);
```
Comment (Krishna2323) [inline src/pages/ReportDetailsPage.tsx:831]: @wildan-m @alitoshmatov is there any specific reason for calling `deleteTransaction` when the component is unmounting? It's causing a crash when deleting a transaction. 

More details here: https://github.com/Expensify/App/pull/66891#issuecomment-3194550954

Can't we use `InteractionManager.runAfterInteractions` to call `deleteTransaction()`?

Comment (wildan-m) [reply src/pages/ReportDetailsPage.tsx:831]: @Krishna2323 did you mean after the component unmounted? At the time of creating this PR, using only `InteractionManager.runAfterInteractions` would still briefly display the not found page. The solution in this PR does not result in any crashes; however, there may be upcoming changes that render it incompatible with the current version.

Comment (Krishna2323) [reply src/pages/ReportDetailsPage.tsx:831]: Yes—at the time of this PR we weren’t doing any state updates, but later we added `removeTransaction`, which performs a state update and led to the crash.

Thanks for the info. I meant calling `InteractionManager.runAfterInteractions` in the callback passed to `onConfirm`, which ensures the state update runs while the component is still mounted and after the animation finishes. On web, I don’t see a “Not Found” page, but I’ll try on mWeb again to see if there’s any problem with the `InteractionManager.runAfterInteractions` approach.

File: src/libs/actions/IOU.ts [hunk 13]
```diff
@@ -5523,10 +5534,59 @@ function prepareToCleanUpMoneyRequest(transactionID: string, reportAction: OnyxT
...
+    // If not a self DM, handle it as a regular money request
+    if (!ReportUtils.isSelfDM(chatReport)) {
+        return getNavigationUrlOnMoneyRequestDelete(transactionID, reportAction, isSingleTransactionView);
+    }
+
+    const transactionThreadID = reportAction.childReportID;
+    const shouldDeleteTransactionThread = transactionThreadID ? (reportAction?.childVisibleActionCount ?? 0) === 0 : false;
+
+    // Only navigate if in single transaction view and the thread will be deleted
+    if (isSingleTransactionView && shouldDeleteTransactionThread && chatReport?.reportID) {
```
Comment (jjcoffee) [inline src/libs/actions/IOU.ts:5582]: Coming from https://github.com/Expensify/App/issues/68735. `shouldDeleteTransactionThread` is unnecessary here, as we always want to navigate after deleting for a self-DM.

File: src/pages/ReportDetailsPage.tsx (hunk 12 above)
Comment (Ollyws) [reply src/pages/ReportDetailsPage.tsx:831]: Just adding in here for checklist, we fixed this in https://github.com/Expensify/App/pull/66891
